x/leverage/keeper/keeper.go

```diff
@@ -271,3 +271,166 @@ func (k Keeper) GetCollateralSetting(ctx sdk.Context, borrowerAddr sdk.AccAddres
+	if err != nil {
+		return err
+	}
+
+	// Use collateral weights to compute borrow limit from enabled collateral
+	borrowLimit, err := k.CalculateBorrowLimit(ctx, collateral)
+	if err != nil {
+		return err
+	}
+
+	// Confirm borrower's eligibility for liquidation
+	if borrowLimit.GTE(borrowValue) {
+		return sdkerrors.Wrap(types.ErrLiquidationIneligible, borrowerAddr.String())
+	}
+
+	// Get reward-specific incentive and dynamic close factor
+	baseRewardDenom := k.FromUTokenToTokenDenom(ctx, rewardDenom)
+	liquidationIncentive, closeFactor, err := k.LiquidationParams(ctx, baseRewardDenom, borrowValue, borrowLimit)
+	if err != nil {
+		return err
+	}
+
+	// Repayment cannot exceed liquidator's available balance
+	liquidatorBalance := k.bankKeeper.GetBalance(ctx, liquidatorAddr, repayment.Denom)
+	if repayment.Amount.GTE(liquidatorBalance.Amount) {
+		repayment.Amount = liquidatorBalance.Amount
+	}
+
+	// Repayment cannot exceed borrower's borrowed amount of selected denom
+	if repayment.Amount.GTE(borrowed.AmountOf(repayment.Denom)) {
+		repayment.Amount = borrowed.AmountOf(repayment.Denom)
+	}
+
+	// Repayment cannot exceed borrowed value * close factor
+	repayValue, err := k.Price(ctx, repayment)
+	if err != nil {
+		return err
+	}
+	if repayValue.GTE(borrowValue.Mul(closeFactor)) {
+		maxRepayValue := borrowValue.Mul(closeFactor)
+		// repayment *= (maxRepayValue / repayValue)
+		repayment.Amount = repayment.Amount.ToDec().Mul(maxRepayValue).Quo(repayValue).TruncateInt()
+		repayValue = maxRepayValue
+	}
+
+	// Given repay denom and amount, use oracle to find equivalent amount of rewardDenom's base asset
+	baseReward, err := k.EquivalentValue(ctx, repayment, baseRewardDenom)
+	if err != nil {
+		return err
+	}
+
+	// Convert reward tokens back to uTokens
+	reward, err := k.ExchangeTokens(ctx, baseReward)
+	if err != nil {
+		return err
+	}
+
+	// Apply liquidation incentive
+	reward.Amount = reward.Amount.ToDec().Mul(sdk.OneDec().Add(liquidationIncentive)).TruncateInt()
+
+	// Reward amount cannot exceed available collateral
+	if reward.Amount.GTE(collateral.AmountOf(rewardDenom)) {
+		// only pay what can be correctly compensated
+		repayment.Amount = repayment.Amount.Mul(collateral.AmountOf(rewardDenom)).Quo(reward.Amount)
+		// use all collateral of selected denom
+		reward.Amount = collateral.AmountOf(rewardDenom)
+	}
+
+	// Send repayment to leverage module account
+	if err := k.bankKeeper.SendCoinsFromAccountToModule(
+		ctx, liquidatorAddr,
+		types.ModuleName,
+		sdk.NewCoins(repayment),
+	); err != nil {
+		return err
+	}
+
+	// Store the remaining borrowed amount in keeper
+	owed := borrowed.AmountOf(repayment.Denom).Sub(repayment.Amount)
+	store := ctx.KVStore(k.storeKey)
+	key := types.CreateLoanKey(borrowerAddr, repayment.Denom)
+	if owed.IsZero() {
+		store.Delete(key) // Completely repaid
+	} else {
+		bz, err := owed.Marshal()
+		if err != nil {
+			return err
+		}
+		store.Set(key, bz) // Partially repaid
+	}
```

Comment:
We should move this entire thing into a keeper method, e.g. `UpdateBorrowed` or something

Reply:
Done, just had to make my existing `SetBorrow` handle the `amount == 0` case and then use it.